auto-value-gson/src/main/java/com/ryanharter/auto/value/gson/AutoValueGsonExtension.java

```diff
@@ -83,12 +84,60 @@
           .addMember("comments", "$S", GENERATED_COMMENTS)
           .build();
 
-  private static final ImmutableSet<TypeName> OPTIONAL_RAW_TYPENAMES = ImmutableSet.of(
-      TypeName.get(com.google.common.base.Optional.class),
-      TypeName.get(java.util.Optional.class),
-      TypeName.get(java.util.OptionalInt.class),
-      TypeName.get(java.util.OptionalDouble.class),
-      TypeName.get(java.util.OptionalLong.class));
```

Comment:
In the first commit, there was a lot of duplication of processing logic to determine how to treat an optional property. This class consolidates that logic in the static factory method `fromPropertyType`.